extensions/client/firebase.go

```diff
@@ -0,0 +1,131 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	firebase "firebase.google.com/go/v4"
+	"firebase.google.com/go/v4/messaging"
+	"github.com/sirupsen/logrus"
+	"github.com/topfreegames/pusher/interfaces"
+	"google.golang.org/api/option"
+	"time"
+)
+
+type firebaseClientImpl struct {
+	firebase *messaging.Client
+	logger   *logrus.Logger
+}
+
+var _ interfaces.PushClient = &firebaseClientImpl{}
+
+func NewFirebaseClient(jsonCredentials string, logger *logrus.Logger) (interfaces.PushClient, error) {
+	ctx := context.Background()
+	projectID, err := getProjectIDFromJson(jsonCredentials)
+	if err != nil {
+		return nil, err
+	}
+	cfg := &firebase.Config{
+		ProjectID: projectID,
+	}
+	app, err := firebase.NewApp(ctx, cfg, option.WithCredentialsJSON([]byte(jsonCredentials)))
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := app.Messaging(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	l := logger.WithFields(logrus.Fields{
+		"source": "firebaseClient",
+	})
+	return &firebaseClientImpl{
+		firebase: client,
+		logger:   l.Logger,
+	}, nil
+}
+
+func (f *firebaseClientImpl) SendPush(ctx context.Context, msg interfaces.Message) error {
+	l := f.logger.WithFields(logrus.Fields{
+		"method": "SendPush",
+	})
+
+	firebaseMsg := toFirebaseMessage(msg)
+	res, err := f.firebase.Send(ctx, &firebaseMsg)
+	if err != nil {
+		l.WithError(err).Error("error sending message")
+		return translateError(err)
+	}
+
+	l.Debugf("Successfully sent message: %s", res)
+
+	return nil
+}
+
+func getProjectIDFromJson(jsonStr string) (string, error) {
+	var data map[string]interface{}
+	err := json.Unmarshal([]byte(jsonStr), &data)
+	if err != nil {
+		return "", err
+	}
+
+	return data["project_id"].(string), nil
```

Comment:
I am not sure if this would panic if "project_id" is not present. Can you confirm, please?

Reply:
Added a check to see if it exists